### PR TITLE
Build Mindustry from source

### PR DESCRIPTION
**Pull request by TobTobXX**
An effort to try to build mindustry from source.

There are still some rough edges (eg. some dependencies have to be manually specified in the generator, support for `x-checker-data`, ...) and this is why it's a WIP PR.

TODOs:
 - [x] remove most of the manually specified deps
 - [x] remove custom build badge
 - [x] generator documentation
 - [x] x-checker-data

### CODE REVIEW

**Comment by TobTobXX** [comment]
Yeah, dunno why it hangs instead of aborts...

Did the jar even work on arm?

Also, how do you start a build job?

**Comment by TheEvilSkeleton** [comment]
> Also, how do you start a build job?

You type in `bot, build` in the comments.

**Comment by TobTobXX** [comment]
> Mindustry doesn't have any native libraries (SDL, OpenAL) compiled for ARM, which is what you see when looking further down the stack trace.
>
> ARM is not supported.

https://github.com/Anuken/Mindustry/issues/2167#issuecomment-644324146

I think arm should be dropped, the jar wouldn't have worked either, probably.

**Comment by TheEvilSkeleton** [comment]
I think it'd be best to drop arm64.

**Comment by TobTobXX** [comment]
bot, build

**Comment by TheEvilSkeleton** [comment]
It works!

I'm not sure, but is the following image here to stay or is it possible to make it say the usual "pre-alpha build 135"?

**Comment by TobTobXX** [comment]
Yeah, I just realized too, joining a server is practically impossible with a from-source built client...

I'll see if there's a way to sign it or something...

But I just read up on `x-data-checker`. It is not needed anymore since it doesn't depend on external data, right?

**Comment by TheEvilSkeleton** [comment]
> But I just read up on `x-data-checker`. It is not needed anymore since it doesn't depend on external data, right?

For it to automatically update to bump Mindustry versions, it's needed.

**Comment by TobTobXX** [comment]
bot, build

**Comment by TobTobXX** [comment]
Screenshot from 2022-02-15 19-17-49

That seems to be fixed. Only `x-checker-data` remaining.

File: com.github.Anuken.Mindustry.yml [hunk 1]
```diff
@@ -18,43 +18,48 @@ finish-args:
...
-      - install -D com.github.Anuken.Mindustry.desktop /app/share/applications/com.github.Anuken.Mindustry.desktop
-      - install -D com.github.Anuken.Mindustry.metainfo.xml /app/share/metainfo/com.github.Anuken.Mindustry.metainfo.xml
-      - install -D com.github.Anuken.Mindustry.png /app/share/icons/hicolor/512x512/apps/com.github.Anuken.Mindustry.png
-
     sources:
+      # Main repo
+      - type: git
+        url: 'https://github.com/Anuken/Mindustry'
+        tag: 'v135'
+        commit: c24c5c6196cd89bdc17a079a2a8acd372cc260d9
```
Comment (TheEvilSkeleton) [inline com.github.Anuken.Mindustry.yml:35]: This should enable x-checker-data.

```suggestion
        commit: c24c5c6196cd89bdc17a079a2a8acd372cc260d9
        x-checker-data:
          type: git
          tag-pattern: "^v([\\d.]+)$"
```

Can confirm it works:

```
TheEvilSkeleton@TheMainDesktop ~/P/com.github.Anuken.Mindustry (beta)> flatpak-external-data-checker com.github.Anuken.Mindustry.yml
INFO    src.manifest: Checking 2 external data items
INFO    src.manifest: Skipped check [1/2] file com.github.Anuken.Mindustry.png (from com.github.Anuken.Mindustry.yml)
INFO    src.manifest: Started check [2/2] git Mindustry (from com.github.Anuken.Mindustry.yml)
INFO    src.lib.externaldata: Source Mindustry: got new version 135
INFO    src.manifest: Finished check [2/2] git Mindustry (from com.github.Anuken.Mindustry.yml)
CHANGE SOON: Mindustry
 Has a new version:
  URL:       https://github.com/Anuken/Mindustry
  Commit:    c24c5c6196cd89bdc17a079a2a8acd372cc260d9
  Tag:       v135
  Branch:    None
  Version:   135
  Timestamp: None
```

**Review by TheEvilSkeleton — CHANGES_REQUESTED** [review]

File: com.github.Anuken.Mindustry.yml (hunk 1 above)
Comment (TheEvilSkeleton) [reply com.github.Anuken.Mindustry.yml:35]: Oops :sweat_smile: https://github.com/flathub/com.github.Anuken.Mindustry/pull/39/commits/1a6417f3595bd16697984617e72d7167fb8dc038

Comment (TobTobXX) [reply com.github.Anuken.Mindustry.yml:35]: Oh, yeah makes sense.

Comment (TobTobXX) [reply com.github.Anuken.Mindustry.yml:35]: Should I remove the `tag:` too?

Comment (TheEvilSkeleton) [reply com.github.Anuken.Mindustry.yml:35]: I prefer if we keep it for semantics.

File: com.github.Anuken.Mindustry.yml [hunk 2]
```diff
@@ -32,7 +32,6 @@ modules:
       - type: git
         url: 'https://github.com/Anuken/Mindustry'
         tag: 'v135'
-        commit: c24c5c6196cd89bdc17a079a2a8acd372cc260d9
```
Comment (TheEvilSkeleton) [inline com.github.Anuken.Mindustry.yml]: Add this again, for reproducibility.

Comment (TobTobXX) [reply com.github.Anuken.Mindustry.yml]: Didn't you just tell me to remove it?

Comment (TheEvilSkeleton) [reply com.github.Anuken.Mindustry.yml]: No, it's just GitHub being GitHub. We can see in the diff https://github.com/flathub/com.github.Anuken.Mindustry/pull/39#discussion_r807190813 that I re-add it. That's just how GitHub works.

**Comment by TobTobXX** [comment]
Oh, and btw, building `stable` is a PITA, because it requires several backporting patches to get a compatible JDK + gradle version. So it'd probably be best if the building manifest keeps in beta until next stable (probably 7.0).

**Comment by TheEvilSkeleton** [comment]
It works really well. Thanks a lot for the docs too.

**Comment by TobTobXX** [comment]
np, it was fun! thank you for being responsive + open :grin:

**Pull request merged**